tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,163 @@
+        "data": lambda data_bytes: int.from_bytes(data_bytes, "big")
+    }
+    return samples, decoders
+
+
+def write_synthetic_streaming_dataset(samples: List[Dict[str, bytes]], shard_size_limit: int) -> str:
+    tmpdir = tempfile.TemporaryDirectory().name
+    first_sample_fields = list(samples[0].keys())
+    with StreamingDatasetWriter(dirname=tmpdir, fields=first_sample_fields,
+                                shard_size_limit=shard_size_limit) as writer:
+        writer.write_samples(samples=samples)
+    return tmpdir
+
+
+@pytest.mark.parametrize("num_samples", [100, 10000])
+@pytest.mark.parametrize("shard_size_limit", [1 << 8, 1 << 16, 1 << 24])
+def test_writer(num_samples: int, shard_size_limit: int) -> None:
+    samples, _ = get_fake_samples_decoders(num_samples)
+
+    first_sample_values = samples[0].values()
+    first_sample_byte_sizes = np.array([len(v) for v in first_sample_values], dtype=np.int64)
+    first_sample_bytes = len(first_sample_byte_sizes.tobytes() + b''.join(first_sample_values))
+
+    expected_samples_per_shard = shard_size_limit // first_sample_bytes
+    expected_num_shards = math.ceil(num_samples / expected_samples_per_shard)
+    expected_num_files = expected_num_shards + 1  # the index file
+
+    dirname = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    files = os.listdir(dirname)
+
+    assert len(files) == expected_num_files, f"Files written ({len(files)}) != expected ({expected_num_files})."
+
+
+@pytest.mark.parametrize("share_remote_local", [False, True])
+@pytest.mark.parametrize("shuffle", [False, True])
+def test_reader(share_remote_local: bool, shuffle: bool):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = tempfile.TemporaryDirectory().name
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=shuffle, decoders=decoders)
+
+    # Test length
+    assert len(dataset) == num_samples, f"Got dataset length={len(dataset)} samples, expected {num_samples}"
+
+    # Test basic sample order
+    shuffle_matches = 0
+    for ix, sample in enumerate(dataset):
+        uid = sample["uid"]
+        data = sample["data"]
+        expected_uid = f"{ix:06}"
+        expected_data = 3 * ix
+        if shuffle:
+            shuffle_matches += (expected_uid == uid)
+        else:
+            assert uid == expected_uid == uid, f"sample ix={ix} has uid={uid}, expected {expected_uid}"
+            assert data == expected_data, f"sample ix={ix} has data={data}, expected {expected_data}"
+
+    # If shuffling, there should be few matches
+    # The probability of k matches in a random permutation is ~1/(e*(k!))
+    if shuffle:
+        assert shuffle_matches < 10
```

Comment:
I presume if shuffling then the sample numbers should not be monotonically increasing? Would it work to validate that invariant directly? This is clever, but a bit hard to follow...

Reply:
I think the non-monotonicity is a weaker check than this one (a reverse order or other poor shuffles would pass it. The `shuffle_matches` is based on a bit of fancy math: the expected number of permutation matches is independent of N (for large enough N) and `1/(e * 10!)` is about 1 in 10million I think it will be pretty safe?

@anisehsani has been working on improved shuffle tests that we may add in afterwards.

Reply:
Sure, not disagreeing that the math might be fancy and correct, but it's not intuitive to follow (where does that equation come from?). Mainly don't want flaky tests from this.